### PR TITLE
[Chuong] search function

### PR DESCRIPTION
**Pull request by onstonboy**
completed

### CODE REVIEW

File: app/src/main/java/com/fstyle/androidtrainning/screen/main/MainActivity.java [hunk 1]
```diff
@@ -8,12 +8,17 @@
 import android.view.Menu;
 import android.view.MenuItem;
 import com.fstyle.androidtrainning.R;
+import com.fstyle.androidtrainning.application.MainApplication;
+import com.fstyle.androidtrainning.data.remote.service.config.LastFmApi;
+import com.fstyle.androidtrainning.model.Track;
```
Comment (daolq3012) [inline app/src/main/java/com/fstyle/androidtrainning/screen/main/MainActivity.java]: ERROR: Unused import: com.fstyle.androidtrainning.model.Track. Use Ctrl+Shift+O (Eclipse) or Ctrl+Alt+O (Intellij) to sort imports

File: app/src/main/java/com/fstyle/androidtrainning/screen/main/MainActivity.java [hunk 2]
```diff
@@ -8,12 +8,17 @@
...
 import com.fstyle.androidtrainning.screen.BaseActivity;
 import com.roughike.bottombar.BottomBar;
 import com.roughike.bottombar.OnTabSelectListener;
+import java.util.List;
```
Comment (daolq3012) [inline app/src/main/java/com/fstyle/androidtrainning/screen/main/MainActivity.java]: ERROR: Unused import: java.util.List. Use Ctrl+Shift+O (Eclipse) or Ctrl+Alt+O (Intellij) to sort imports

File: app/src/main/java/com/fstyle/androidtrainning/screen/main/MainContract.java [hunk 3]
```diff
@@ -1,5 +1,9 @@
 package com.fstyle.androidtrainning.screen.main;
 
+import com.fstyle.androidtrainning.data.remote.service.config.LastFmApi;
+import com.fstyle.androidtrainning.model.Track;
```
Comment (daolq3012) [inline app/src/main/java/com/fstyle/androidtrainning/screen/main/MainContract.java]: ERROR: Unused import: com.fstyle.androidtrainning.model.Track. Use Ctrl+Shift+O (Eclipse) or Ctrl+Alt+O (Intellij) to sort imports

File: app/src/main/java/com/fstyle/androidtrainning/screen/main/MainContract.java [hunk 4]
```diff
@@ -1,5 +1,9 @@
...
+import java.util.List;
```
Comment (daolq3012) [inline app/src/main/java/com/fstyle/androidtrainning/screen/main/MainContract.java]: ERROR: Unused import: java.util.List. Use Ctrl+Shift+O (Eclipse) or Ctrl+Alt+O (Intellij) to sort imports

File: app/src/main/java/com/fstyle/androidtrainning/screen/main/MainActivity.java (hunk 1 above)
Comment (daolq3012) [inline app/src/main/java/com/fstyle/androidtrainning/screen/main/MainActivity.java]: WARNING: [UnusedImports] Avoid unused imports such as 'com.fstyle.androidtrainning.model.Track'
https://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/imports.html#UnusedImports

File: app/src/main/java/com/fstyle/androidtrainning/screen/main/MainActivity.java (hunk 2 above)
Comment (daolq3012) [inline app/src/main/java/com/fstyle/androidtrainning/screen/main/MainActivity.java]: WARNING: [UnusedImports] Avoid unused imports such as 'java.util.List'
https://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/imports.html#UnusedImports

File: app/src/main/java/com/fstyle/androidtrainning/screen/main/MainContract.java (hunk 3 above)
Comment (daolq3012) [inline app/src/main/java/com/fstyle/androidtrainning/screen/main/MainContract.java]: WARNING: [UnusedImports] Avoid unused imports such as 'com.fstyle.androidtrainning.model.Track'
https://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/imports.html#UnusedImports

File: app/src/main/java/com/fstyle/androidtrainning/screen/main/MainContract.java (hunk 4 above)
Comment (daolq3012) [inline app/src/main/java/com/fstyle/androidtrainning/screen/main/MainContract.java]: WARNING: [UnusedImports] Avoid unused imports such as 'java.util.List'
https://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/imports.html#UnusedImports

File: app/src/main/java/com/fstyle/androidtrainning/model/AlbumResults.java [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+    @SerializedName("opensearch:startIndex")
+    @Expose
+    private String opensearchStartIndex;
+    @SerializedName("opensearch:itemsPerPage")
+    @Expose
+    private String opensearchItemsPerPage;
+    @SerializedName("albummatches")
+    @Expose
+    private Albummatches albummatches;
+    @SerializedName("@attr")
```
Comment (daolq3012) [inline app/src/main/java/com/fstyle/androidtrainning/model/AlbumResults.java:26]: ERROR: [PMD-CPD] 37 lines duplicated.
app/src/main/java/com/fstyle/androidtrainning/model/TrackResults.java

File: app/src/main/java/com/fstyle/androidtrainning/model/TrackResults.java [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+    @SerializedName("opensearch:startIndex")
+    @Expose
+    private String opensearchStartIndex;
+    @SerializedName("opensearch:itemsPerPage")
+    @Expose
+    private String opensearchItemsPerPage;
+    @SerializedName("trackmatches")
+    @Expose
+    private Trackmatches trackmatches;
+    @SerializedName("@attr")
```
Comment (daolq3012) [inline app/src/main/java/com/fstyle/androidtrainning/model/TrackResults.java:27]: ERROR: [PMD-CPD] 37 lines duplicated.
app/src/main/java/com/fstyle/androidtrainning/model/AlbumResults.java

File: app/src/main/java/com/fstyle/androidtrainning/screen/main/MainPresenter.java [hunk 7]
```diff
@@ -22,4 +36,67 @@ public void onStart() {
...
+    private void doSearchAlbum(String newText) {
+        mApi.getAlbumSearch(Constant.METHOD_SEARCH_ALBUM, newText, Constant.API_KEY,
+                Constant.FORMAT_TYPE_API).enqueue(new Callback<AlbumSearchResponse>() {
+            @Override
+            public void onResponse(@NonNull Call<AlbumSearchResponse> call,
+                    @NonNull Response<AlbumSearchResponse> response) {
+                if (response.body().getResults() != null
+                        && response.body().getResults().getAlbummatches().getAlbum() != null) {
+                    for (int i = MIN; i < SEARCH_AMOUNT; i++) {
+                        Log.d(TAG, "doSearchAlbum: " + response.body()
```
Comment (tranminhtri911) [inline app/src/main/java/com/fstyle/androidtrainning/screen/main/MainPresenter.java]: remove log di em, dặn bao lần dòi (kill)

File: app/src/main/java/com/fstyle/androidtrainning/screen/main/MainPresenter.java [hunk 8]
```diff
@@ -22,4 +36,67 @@ public void onStart() {
...
+
+    private void doSearchAlbum(String newText) {
+        mApi.getAlbumSearch(Constant.METHOD_SEARCH_ALBUM, newText, Constant.API_KEY,
+                Constant.FORMAT_TYPE_API).enqueue(new Callback<AlbumSearchResponse>() {
+            @Override
+            public void onResponse(@NonNull Call<AlbumSearchResponse> call,
+                    @NonNull Response<AlbumSearchResponse> response) {
+                if (response.body().getResults() != null
+                        && response.body().getResults().getAlbummatches().getAlbum() != null) {
+                    for (int i = MIN; i < SEARCH_AMOUNT; i++) {
```
Comment (tranminhtri911) [inline app/src/main/java/com/fstyle/androidtrainning/screen/main/MainPresenter.java]: remove MIn thay = 0 đi em

File: app/src/main/java/com/fstyle/androidtrainning/screen/main/MainPresenter.java [hunk 9]
```diff
@@ -1,14 +1,138 @@
...
+                Constant.FORMAT_TYPE_API).enqueue(new Callback<ArtistSearchResponse>() {
+            @Override
+            public void onResponse(@NonNull Call<ArtistSearchResponse> call,
+                    @NonNull Response<ArtistSearchResponse> response) {
+                if (response.body() == null
+                        || response.body().getResults().getArtistmatches().getArtist() == null) {
+                    isFetchDataSuccess = false;
+                    return;
+                }
+                if (response.body().getResults().getArtistmatches().getArtist().isEmpty()){
```
Comment (daolq3012) [inline app/src/main/java/com/fstyle/androidtrainning/screen/main/MainPresenter.java]: ERROR: '{' is not preceded with whitespace.

File: app/src/main/java/com/fstyle/androidtrainning/screen/main/MainPresenter.java [hunk 10]
```diff
@@ -1,14 +1,138 @@
...
+                Constant.FORMAT_TYPE_API).enqueue(new Callback<AlbumSearchResponse>() {
+            @Override
+            public void onResponse(@NonNull Call<AlbumSearchResponse> call,
+                    @NonNull Response<AlbumSearchResponse> response) {
+                if (response.body() == null
+                        || response.body().getResults().getAlbummatches().getAlbum() == null) {
+                    isFetchDataSuccess = false;
+                    return;
+                }
+                if (response.body().getResults().getAlbummatches().getAlbum().isEmpty()){
```
Comment (daolq3012) [inline app/src/main/java/com/fstyle/androidtrainning/screen/main/MainPresenter.java]: ERROR: '{' is not preceded with whitespace.

File: app/src/main/java/com/fstyle/androidtrainning/screen/main/MainPresenter.java [hunk 11]
```diff
@@ -1,14 +1,138 @@
...
+                Constant.FORMAT_TYPE_API).enqueue(new Callback<TrackSearchResponse>() {
+            @Override
+            public void onResponse(@NonNull Call<TrackSearchResponse> call,
+                    @NonNull Response<TrackSearchResponse> response) {
+                if (response.body() == null
+                        || response.body().getResults().getTrackmatches().getTrack() == null) {
+                    isFetchDataSuccess = false;
+                    return;
+                }
+                if (response.body().getResults().getTrackmatches().getTrack().isEmpty()){
```
Comment (daolq3012) [inline app/src/main/java/com/fstyle/androidtrainning/screen/main/MainPresenter.java]: ERROR: '{' is not preceded with whitespace.

File: app/src/main/java/com/fstyle/androidtrainning/screen/main/SearchSongRecyclerAdapter.java [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+            String urlImage = mTracks.get(position).getImage().get(MEDIUM_IMAGE).getText();
+            if (!urlImage.isEmpty()) {
+                Glide.with(mContext).load(urlImage).into(mImageView);
+            } else {
+                mImageView.setImageResource(R.drawable.img_demo_50x50);
+            }
+        }
+
+        private void setNameArtist(int position) {
+            String nameArtist = mTracks.get(position).getArtist();
```
Comment (daolq3012) [inline app/src/main/java/com/fstyle/androidtrainning/screen/main/SearchSongRecyclerAdapter.java:94]: ERROR: [PMD-CPD] 20 lines duplicated.
app/src/main/java/com/fstyle/androidtrainning/screen/main/mainfragments/online/subfragment/toptrack/TopTrackRecyclerAdapter.java

File: app/src/main/java/com/fstyle/androidtrainning/model/AlbumResults.java (hunk 5 above)
Comment (daolq3012) [inline app/src/main/java/com/fstyle/androidtrainning/model/AlbumResults.java:26]: ERROR: [PMD-CPD] 37 lines duplicated.
app/src/main/java/com/fstyle/androidtrainning/model/ArtistResult.java
app/src/main/java/com/fstyle/androidtrainning/model/TrackResults.java

File: app/src/main/java/com/fstyle/androidtrainning/model/ArtistResult.java [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+    @SerializedName("opensearch:startIndex")
+    @Expose
+    private String opensearchStartIndex;
+    @SerializedName("opensearch:itemsPerPage")
+    @Expose
+    private String opensearchItemsPerPage;
+    @SerializedName("artistmatches")
+    @Expose
+    private Artistmatches artistmatches;
+    @SerializedName("@attr")
```
Comment (daolq3012) [inline app/src/main/java/com/fstyle/androidtrainning/model/ArtistResult.java:26]: ERROR: [PMD-CPD] 37 lines duplicated.
app/src/main/java/com/fstyle/androidtrainning/model/AlbumResults.java
app/src/main/java/com/fstyle/androidtrainning/model/TrackResults.java

File: app/src/main/java/com/fstyle/androidtrainning/model/TrackResults.java (hunk 6 above)
Comment (daolq3012) [inline app/src/main/java/com/fstyle/androidtrainning/model/TrackResults.java:27]: ERROR: [PMD-CPD] 37 lines duplicated.
app/src/main/java/com/fstyle/androidtrainning/model/AlbumResults.java
app/src/main/java/com/fstyle/androidtrainning/model/ArtistResult.java

**Comment by daolq3012** [comment]
OK!

**Pull request merged**